codec-http/src/test/java/io/netty/handler/codec/http/websocketx/extensions/compression/PerMessageDeflateDecoderTest.java

```diff
@@ -332,54 +358,86 @@ public void testFragmentedFrameWithLeftOverInLastFragment() {
                          "f6f736e73746575637a6d727a7175707a6e74627578687871767771697a71766c64626d78726d6d7675756877" +
                          "62667963626b687a726d676e646263776e67797264706d6c6863626577616967706a78636a72697464756e627" +
                          "977616f79736475676f76736f7178746a7a7479626c64636b6b6778637768746c62";
-        EmbeddedChannel encoderChannel = new EmbeddedChannel(
-                ZlibCodecFactory.newZlibEncoder(ZlibWrapper.NONE, 9, 15, 8));
+        EmbeddedChannel encoderChannel = new EmbeddedChannel(BYTEBUF_TO_BUFFER_HANDLER,
+                ZlibCodecFactory.newZlibEncoder(ZlibWrapper.NONE, 9, 15, 8),
+                BUFFER_TO_BYTEBUF_HANDLER);
         EmbeddedChannel decoderChannel = new EmbeddedChannel(new PerMessageDeflateDecoder(false));
 
-        ByteBuf originPayload = Unpooled.wrappedBuffer(ByteBufUtil.decodeHexDump(hexDump));
-        assertTrue(encoderChannel.writeOutbound(originPayload.duplicate().retain()));
-
-        ByteBuf compressedPayload = encoderChannel.readOutbound();
-        compressedPayload = compressedPayload.slice(0, compressedPayload.readableBytes() - 4);
+        Buffer originPayload = encoderChannel.bufferAllocator().copyOf(ByteBufUtil.decodeHexDump(hexDump));
+        assertTrue(encoderChannel.writeOutbound(originPayload.copy()));
 
-        int oneThird = compressedPayload.readableBytes() / 3;
-
-        TextWebSocketFrame compressedFrame1 = new TextWebSocketFrame(
-                false, WebSocketExtension.RSV1, compressedPayload.slice(0, oneThird));
-        ContinuationWebSocketFrame compressedFrame2 = new ContinuationWebSocketFrame(
-                false, WebSocketExtension.RSV3, compressedPayload.slice(oneThird, oneThird));
-        ContinuationWebSocketFrame compressedFrame3 = new ContinuationWebSocketFrame(
-                false, WebSocketExtension.RSV3, compressedPayload.slice(oneThird * 2, oneThird));
-        int offset = oneThird * 3;
-        ContinuationWebSocketFrame compressedFrameWithExtraData = new ContinuationWebSocketFrame(
-                true, WebSocketExtension.RSV3, compressedPayload.slice(offset,
-                     compressedPayload.readableBytes() - offset));
+        Buffer compressedPayload = encoderChannel.readOutbound();
+        TextWebSocketFrame compressedFrame1;
+        ContinuationWebSocketFrame compressedFrame2;
+        ContinuationWebSocketFrame compressedFrame3;
+        ContinuationWebSocketFrame compressedFrameWithExtraData;
+        try {
+            compressedPayload = compressedPayload.readSplit(compressedPayload.readableBytes() - 4);
```

Comment:
We're forgetting to close the remains of the original `compressedPayload`. Maybe `skipWritable(-4)` is better?